File: backend/src/ytclip/domain/timestamps.py
"""Timestamp grammar shared with the UI: `SS`, `MM:SS`, `HH:MM:SS`, `[Nh][Nm][Ns]`."""

import re

_DIGITS_RE = re.compile(r"[0-9]+")
_HMS_RE = re.compile(r"(?:([0-9]+)h)?(?:([0-9]+)m)?(?:([0-9]+)s)?")


def parse_timestamp(value: str | int) -> int:
    """Return whole seconds, or raise ValueError for anything outside the grammar."""
    if isinstance(value, bool):
        raise ValueError("a boolean is not a timestamp")
    if isinstance(value, int):
        if value < 0:
            raise ValueError("timestamps cannot be negative")
        return value

    text = str(value).strip().lower()
    if not text:
        raise ValueError("empty timestamp")
    if _DIGITS_RE.fullmatch(text):
        return int(text)

    if ":" in text:
        parts = text.split(":")
        if len(parts) not in (2, 3) or not all(_DIGITS_RE.fullmatch(part) for part in parts):
            raise ValueError(f"unrecognised timestamp: {value!r}")
        numbers = [int(part) for part in parts]
        if len(numbers) == 3:
            hours, minutes, seconds = numbers
            if minutes > 59 or seconds > 59:
                raise ValueError("minutes and seconds must be below 60 in HH:MM:SS")
            return hours * 3600 + minutes * 60 + seconds
        # MM:SS is deliberately lenient: "1:65" means 125 seconds.
        minutes, seconds = numbers
        return minutes * 60 + seconds

    match = _HMS_RE.fullmatch(text)
    if not match or not any(match.groups()):
        raise ValueError(f"unrecognised timestamp: {value!r}")
    hours, minutes, seconds = (int(group) if group else 0 for group in match.groups())
    return hours * 3600 + minutes * 60 + seconds


def _split(seconds: int) -> tuple[int, int, int]:
    seconds = max(0, int(seconds))
    return seconds // 3600, (seconds % 3600) // 60, seconds % 60


def format_hms(seconds: int) -> str:
    hours, minutes, secs = _split(seconds)
    return f"{hours:02d}:{minutes:02d}:{secs:02d}"


def format_hms_for_filename(seconds: int) -> str:
    hours, minutes, secs = _split(seconds)
    return f"{hours:02d}-{minutes:02d}-{secs:02d}"

File: extension/src/timestamps.js
// Timestamp grammar shared with the backend: `SS`, `MM:SS`, `HH:MM:SS`, `[Nh][Nm][Ns]`.

const DIGITS = /^\d+$/;
const HMS = /^(?:(\d+)h)?(?:(\d+)m)?(?:(\d+)s)?$/;

/** Whole seconds, or throws for anything outside the grammar. */
export function parseTimestamp(value) {
  const text = String(value ?? "").trim().toLowerCase();
  if (!text) throw new Error("Enter a time like 1:30, 90 or 1m30s.");
  if (DIGITS.test(text)) return Number.parseInt(text, 10);
  if (text.includes(":")) {
    const parts = text.split(":");
    if (![2, 3].includes(parts.length) || !parts.every((part) => DIGITS.test(part))) {
      throw new Error(`"${value}" is not a time. Use MM:SS or HH:MM:SS.`);
    }
    const numbers = parts.map((part) => Number.parseInt(part, 10));
    if (numbers.length === 3) {
      const [hours, minutes, seconds] = numbers;
      if (minutes > 59 || seconds > 59) throw new Error("Minutes and seconds must be below 60 in HH:MM:SS.");
      return hours * 3600 + minutes * 60 + seconds;
    }
    const [minutes, seconds] = numbers; // MM:SS is deliberately lenient: "1:65" means 125 seconds.
    return minutes * 60 + seconds;
  }
  const match = HMS.exec(text);
  if (!match || !(match[1] || match[2] || match[3])) throw new Error(`"${value}" is not a time. Use MM:SS, HH:MM:SS or 1h2m3s.`);
  return Number(match[1] || 0) * 3600 + Number(match[2] || 0) * 60 + Number(match[3] || 0);
}

function split(seconds) {
  const total = Math.max(0, Math.trunc(seconds));
  return [Math.floor(total / 3600), Math.floor((total % 3600) / 60), total % 60];
}

const pad = (n) => String(n).padStart(2, "0");

export function formatHms(seconds) {
  const [h, m, s] = split(seconds);
  return `${pad(h)}:${pad(m)}:${pad(s)}`;
}

export function formatHmsForFilename(seconds) {
  const [h, m, s] = split(seconds);
  return `${pad(h)}-${pad(m)}-${pad(s)}`;
}
